[PF-215] Add cache to experiment save call path

amangup opened the pull request:
The experiment save path compares the experiment config in the call with the one stored in redis, and updates it if its different. It makes redis calls in that comparison, which we can avoid by using the cache introduced in our last split change to get experiment config.

Also added clearing of cache on `experiment#reset`. This fixes a small issue where any experiment config changes (like adding a new variant) would need two reboots to go live.

amangup committed:
Cache another redis call path in the experiment.save call

amangup committed:
1) Modify the caching mechanism for checking experiment config change, and 2) Clear cache on the reset path

amangup committed:
Delete an extra line

rdh commented on lib/split/cache.rb:
```diff
@@ -18,5 +18,9 @@ def self.fetch(namespace, key)
+
+    def self.clear_key(namespace, key)
+      @cache[namespace]&.delete(key)
+    end
```
Hmm ... I'm not sure about this.  It will clear the key in the local server, but other servers in the cluster will still have stale cached data.  Think it may be better for all servers to be in the same state, synchronized by restart, rather than having all stale, and one reset.

rdh replied:
I guess, in a low-traffic cluster, some set of servers could cache before a change and others after, so this isn't really a new problem related to this logic.  More of a general local-cache issue.

amangup replied:
We talked about this off line --- all servers will update to the latest value in a couple of requests at most.

rdh replied:
So, I'd like to avoid clearing specific keys.  I think we just `Split::Cache.clear` anytime stuff gets stale, rather than go down the rabbit hole of keeping track of which keys when.  Will be a little bit less efficient, but way less error prone.

rdh replied:
To be explicit, let's get rid of this method, and just invalidate the whole cache anywhere it's needed.

amangup replied:
Is there any specific reason? Within the experiment class, it's not ambiguous which key is being cleared.

amangup replied:
Just to expand on my reasoning for choosing this way --- Cache is a standard concept in computing and there is a certain expected behavior from them. They all have an eviction policy, and it is defined as criterion (key) --> evict(key). In this case we have a explicit staleness criteria. Unless there is a specific usage pattern that is different (which I don't see here), why should we deviate from the norm?

rdh replied:
Ah, mr primary concern with the original implementation/usage is that we'd need maintain the clearing logic, as well.  So if we cached another experiment value, we'd also need to know to add another line to clear it from the cache in the reset logic.  

Looks like your new approach resolves that method, as long as we use the same name for the experiment everywhere.  I'm cool with this new strategy.

amangup committed:
Make the cache#clear_key method clear that key across all namespaces. To make sure we can clean all experiment caching without having to keep track of all the namespaces its details are stored in.

amangup committed:
Forgot a colon in a symbol

amangup committed:
Another small issue with variable using.

amangup committed:
Another typo fix: Split::cache -> Split::Cache

amangup commented:
Instead of clearing the whole cache on experiment reset, I am clearing that experiment from all namespaces. I think it would be unintuitive for the whole cache to be cleared on change for one experiment, and can catch someone out in the future as an unexpected behavior.

The main problem with previous implementation was that there could be many namespaces to keep track of, which is no longer a concern.

amangup committed:
1) Fix potential nil error
2) On experiment reset, make sure before reset hook calls before cache is cleared.

amangup committed:
Make sure that experiment config in Experiment catalog is the same as the one used in Experiment#experiment_configuration_has_changed?

amangup commented:
@rdh Another change related to discussion we had -- I am making sure that `ExperimentCatalog#find` and `experiment_configuration_has_changed?` use exactly the same cached value. Otherwise, it is possible that `experiment_configuration_has_changed?` gets the new config from redis (thus not going for the reset path), while `ExperimentCatalog#find` continues to have the old config -- persisting the issue which prevents cached experiment config to update.

rdh commented on lib/split/experiment_catalog.rb:
```diff
@@ -13,10 +13,7 @@ def self.all_active_first
     end
 
     def self.find(name)
-      Split.cache(:experiment_catalog, name) do
```
I kinda like keeping caching at the Catalog level.  Like the low-level experiment always loads, so you have a path to get the latest, if you really want it.  Like dashboard could use the low-level always load, but there's a higher level that caches for frequent operations.  No idea if that fits the overall arch, but is how I'm thinking about it.

amangup replied:
The functionality is not affected by this change....the change still means `ExperimentCatalog#find` call will be cached. The dashboard's low level call, if it's bypassing `ExperimentCatalog` will still continue to load the uncached version.

There are two ways we can think about it:
- `Experiment` class accesses the directory of experiments (`ExperimentCatalog`) to know what the current experiment config is. I personally feel this is a reasonable relationsip, but you mentioned `Experiment` class using `ExperimentCatalog` feels like child class is calling its parent.
- Add a factory method to `Experiment`, which has caching, and have both `ExperimentCatalog` and `Experiment` class  use this. 

I don't know if there is a third solution....but as I mentioned before, we have to use the same cache namespace to return the experiment config for both calling locations.

rdh replied:
I think the refactor you did of the `find` method down to the `Experiment` model is good, and resolves my concern 👍

rdh reviewed (approved):
👍 

You'll want to use `bundle update split --conservative` in the main repo to update the gem reference to this latest version after merging.

stepchud commented on lib/split/experiment.rb:
```diff
@@ -13,6 +13,13 @@ class Experiment
       :resettable => true
     }
 
+    def self.find(name)
```
we're sure this method name doesn't collide anywhere right?

amangup replied:
Checked...not in the `Split::Experiment` namespace
